feat(sdk): Make created_dm_room public

timthelion opened the pull request:
Fixes #908

jplatte commented on crates/matrix-sdk/src/encryption/mod.rs:
```diff
@@ -230,8 +230,9 @@ impl Client {
-    pub(crate) async fn create_dm_room(
+    pub async fn create_dm_room(
         &self,
         user_id: OwnedUserId,
```
I think we should make this `&UserId` for consistency with other (public) methods. I think the only change required for this internally is changing `user_id.clone()` to `user_id.to_owned()`.

timthelion replied:
Done

poljar reviewed (changes requested):
Before this method can be made public, it should be adapted to utilize what has been merged in #868.

It also doesn't make sense to keep this method feature gated behind the `e2e-encryption` feature.

Optionally, it may also make sense to add a shorthand method that operates on the `RoomMember` struct:
```rust
RoomMember::create_dm(&self) -> Result<Joined>;
RoomMember::get_or_create_dm(&self) -> Result<Joined>;
```

timthelion commented:
@poljar I tried removing the e2e feature flag and found that `create_dm_room` depends on `send_account_data` which is also feature flagged. So I removed the feature flag on that too and moved both to the `client` module from the `encryption` module.

Instead of duplicating code to implement an event handler on the `RoomCreateEventContent` like in `create_room` I delegated that task to the original `create_room` method. I also think that I found a bug in `create_room`. I'm not sure but it seems to me, that if the room already existed, then there would be no event fired, and this would hang forever. So I added some code that looks up the room using `get_joined_room` and returns early if that finds a room in the store. I'm not %100 clear on races, but it seems to me that if that returns early, than the room was "gotten" and not created. Therefore I made a new function `get_or_create_room` which tells us if the room was "gotten" or "created". If it turns out my assumptions are false, we can simply eliminate the new function.

As of now, I have implemented this in a series of commits. Once we have ironed out the details, then I can squash it all.

I understand the desire for the `RoomMember` method, however I don't need this functionality and therefore don't want to implement it.

poljar commented:
> I'm not sure but it seems to me, that if the room already existed, then there would be no event fired, and this would hang forever.

I'm a bit confused, how would a room already exist? Telling the server to create a room always creates a new room with a unique room id. The response you receive contains a new unique room id. 

The only way I can think of that this can happen is if the event handler gets registered after the request is made.

This doesn't seem to be the case though.

timthelion commented:
OK, it wasn't clear to me whether creating a new PM with a person with whom you are already in a PM actually creates a new PM or simply returns the old PM. I guess the second behavior is actually better.

poljar commented:
> OK, it wasn't clear to me whether creating a new PM with a person with whom you are already in a PM actually creates a new PM or simply returns the old PM. I guess the second behavior is actually better.

Oh no, you can actually crate an infinite amount of PMs, as you notice something being a PM/DM is just something you mark yourself with the account data event:
https://github.com/matrix-org/matrix-rust-sdk/pull/961/files#diff-6d2ff673b14b2a563388b2d177b0c4e7ff2c16714054858fdec9bfd01858ba96R1940

A `get_or_create()` method does indeed make sense for DMs though, if you need this you can combine it with the `get_dm_room()` method:
https://github.com/matrix-org/matrix-rust-sdk/blob/f980368740d4a9c356a3d97a4fd96a6aab7557f3/crates/matrix-sdk/src/encryption/mod.rs#L381-L393

After all, having multiple DMs isn't usually desirable.

poljar commented on crates/matrix-sdk/src/client/mod.rs:
````diff
@@ -1793,6 +1796,43 @@ impl Client {
+    /// assert!(client.create_room(request).await.is_ok());
+    /// # });
+    /// ```
+    pub async fn get_or_create_room(
````
As per comment before, I don't think it's possible to get or create a new room. The hang you experienced might be because:
1. A sync loop was not running
2. You called this inside an event handler

If you want, a `get_or_create_dm()` method does make sense, but that's not required to get this merged.

poljar commented on crates/matrix-sdk/src/client/mod.rs:
```diff
@@ -1847,14 +1893,72 @@ impl Client {
     }
 
+
+    async fn send_account_data<T>(
```
This has been made public as part of #964.

poljar commented:
Thanks, this has been merged as part of https://github.com/matrix-org/matrix-rust-sdk/pull/1015.
